contracts/allocators/alchemistAllocator.sol

```diff
@@ -0,0 +1,210 @@
+// SPDX-License-Identifier: AGPL-3.0-or-later
+pragma solidity 0.7.5;
+
+import "../libraries/Address.sol";
+import "../libraries/SafeMath.sol";
+import "../libraries/SafeERC20.sol";
+
+import "../interfaces/IERC20.sol";
+import "../interfaces/IERC20Metadata.sol";
+import "../interfaces/ITreasury.sol";
+
+import "../types/Ownable.sol";
+
+interface ITokemak_Manager {
+    function currentCycleIndex() external view returns (uint256);
+}
+
+interface ITokemak_tALCX {
+    function deposit(uint256 amount) external;
+
+    function requestedWithdrawals(address addr) external view returns (uint256, uint256);
+
+    function withdraw(uint256 requestedAmount) external;
+
+    function requestWithdrawal(uint256 amount) external;
+}
+
+interface IStakingPools {
+    function claim(uint256 _poolId) external;
+
+    function exit(uint256 _poolId) external;
+
+    function getStakeTotalDeposited(address _account, uint256 _poolId) external view returns (uint256);
+
+    function getStakeTotalUnclaimed(address _account, uint256 _poolId) external view returns (uint256);
+
+    function deposit(uint256 _poolId, uint256 _depositAmount) external;
+
+    function withdraw(uint256 _poolId, uint256 _withdrawAmount) external;
+}
+
+/**
+ *  Contract deploys Alchemist from treasury into the Tokemak tALCX pool,
+ *  tALCX contract gives tALCX token in ratio 1:1 of Alchemist token deposited,
+ *  Contract stake tALCX token on Alchemist staking pool and earn ALCX as reward,
+ *  Contract claims reward and compound it,
+ *  Contract withdraw funds from Alchemist staking pool and Tokemak tALCX pool,
+ *  Sends back Alchemist token with accured reward to treasury.
+ */
+
+contract AlchemistAllocator is Ownable {
+    /* ======== DEPENDENCIES ======== */
+
+    using SafeERC20 for IERC20;
+    using SafeMath for uint256;
+
+    /* ======== STATE VARIABLES ======== */
+
+    address alchemist;
+    address tokemak_manager = 0xA86e412109f77c45a3BC1c5870b880492Fb86A14;
+
+    ITokemak_tALCX immutable tALCX; // Tokemak tALCX deposit contract
+    IStakingPools immutable pool; // Alchemist staking contract
+    ITreasury immutable treasury; // Olympus Treasury
+
+    uint256 public totalAlchemistDeposited;
+
+    /* ======== CONSTRUCTOR ======== */
+
+    constructor(
+        address _treasury,
+        address _alchemist,
+        address _tALCX,
+        address _pool
+    ) {
+        require(_treasury != address(0));
+        treasury = ITreasury(_treasury);
+
+        require(_tALCX != address(0));
+        tALCX = ITokemak_tALCX(_tALCX);
+
+        require(_pool != address(0));
+        pool = IStakingPools(_pool);
+
+        alchemist = _alchemist;
+    }
+
+    /* ======== POLICY FUNCTIONS ======== */
+
+    /**
+     *  @notice sets Alchemist address
+     */
+    function setAlchemistAddress(address _alchemist) external onlyOwner {
```

Comment:
Can remove this once `alchemix` is immutable

Reply:
Removed